{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 57101 entries, 0 to 57100\n",
      "Data columns (total 8 columns):\n",
      "CallTime        57101 non-null datetime64[ns]\n",
      "DispatchTime    54286 non-null datetime64[ns]\n",
      "IncidentDesc    57101 non-null object\n",
      "Disposition     57101 non-null object\n",
      "EventNumber     57101 non-null int64\n",
      "Location        57101 non-null object\n",
      "Priority        57101 non-null object\n",
      "Case            19090 non-null object\n",
      "dtypes: datetime64[ns](2), int64(1), object(5)\n",
      "memory usage: 3.5+ MB\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"police_calls_2019.csv\", sep=';', encoding='cp1252')\n",
    "\n",
    "df = df.rename(columns={\"0\":\"x\", \"1\": \"CallTime\", \"2\": \"DispatchTime\", \"3\":\"IncidentDesc\", \"4\": \"Disposition\", \"5\": \"EventNumber\", \"6\":\"Location\", \"7\":\"Priority\", \"8\":\"Case\"})\n",
    "\n",
    "df = df.drop(columns={\"x\"})\n",
    "\n",
    "df['CallTime'] = pd.to_datetime(df['CallTime'])\n",
    "df['DispatchTime'] = pd.to_datetime(df['DispatchTime'])\n",
    "\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "import geopandas as gpd\n",
    "import geopy\n",
    "from geopy.geocoders import Nominatim\n",
    "from geopy.extra.rate_limiter import RateLimiter\n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2920, Olive Street, Woodfield Microhood, College Hill, Eugene, Lane County, Oregon, 97405, United States of America\n",
      "Latitude = 44.02567120249996, Longitude = -123.095035\n"
     ]
    }
   ],
   "source": [
    "locator = Nominatim(user_agent=\"myGeocoder\")\n",
    "location = locator.geocode(\"2920 Olive ST,Eugene,Oregon\")\n",
    "print(location.address)\n",
    "print(\"Latitude = {}, Longitude = {}\".format(location.latitude, location.longitude))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'address_components': [{'long_name': 'Franklin Boulevard & Alder Street', 'short_name': 'Franklin Blvd & Alder St', 'types': ['intersection']}, {'long_name': 'Eugene', 'short_name': 'Eugene', 'types': ['locality', 'political']}, {'long_name': 'Lane County', 'short_name': 'Lane County', 'types': ['administrative_area_level_2', 'political']}, {'long_name': 'Oregon', 'short_name': 'OR', 'types': ['administrative_area_level_1', 'political']}, {'long_name': 'United States', 'short_name': 'US', 'types': ['country', 'political']}, {'long_name': '97401', 'short_name': '97401', 'types': ['postal_code']}], 'formatted_address': 'Franklin Blvd & Alder St, Eugene, OR 97401, USA', 'geometry': {'location': {'lat': 44.0496119, 'lng': -123.0802335}, 'location_type': 'GEOMETRIC_CENTER', 'viewport': {'northeast': {'lat': 44.0509608802915, 'lng': -123.0788845197085}, 'southwest': {'lat': 44.0482629197085, 'lng': -123.0815824802915}}}, 'place_id': 'Ei9GcmFua2xpbiBCbHZkICYgQWxkZXIgU3QsIEV1Z2VuZSwgT1IgOTc0MDEsIFVTQSJmImQKFAoSCZ0SzYY9HsFUEWDl-TnBBWkSEhQKEgmdEs2GPR7BVBFg5fk5wQVpEhoUChIJiW1HmI0fwVQRHxzATE6zj4IaFAoSCdej_9myH8FUESkkCcGh48s-IgoN929BGhVhcqO2', 'types': ['intersection']}]\n"
     ]
    }
   ],
   "source": [
    "import googlemaps\n",
    "gmaps = googlemaps.Client(key='AIzaSyAC56-T3bMZdV_59Ki7p_xqR7fSRSky_xU')\n",
    "\n",
    "# Geocoding an address\n",
    "geocode_result = gmaps.geocode('ALDER ST/FRANKLIN BLVD, Eugene, OR')\n",
    "print(geocode_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(geocode_result[0]['geometry']['location']['lat'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 CallTime        DispatchTime             IncidentDesc  \\\n",
      "22034 2019-06-01 23:59:37 2019-06-02 01:20:52        Criminal Trespass   \n",
      "22035 2019-06-01 23:56:05 2019-06-02 00:04:52  Suspicious Condition(s)   \n",
      "22036 2019-06-01 23:46:29 2019-06-01 23:46:29             Traffic Stop   \n",
      "22037 2019-06-01 23:41:23 2019-06-02 04:30:48         Animal Complaint   \n",
      "22038 2019-06-01 23:35:05 2019-06-02 00:17:35  Suspicious Condition(s)   \n",
      "...                   ...                 ...                      ...   \n",
      "48570 2019-11-08 00:45:11 2019-11-08 00:45:11              Person Stop   \n",
      "48571 2019-11-08 00:24:11 2019-11-08 00:24:11              Person Stop   \n",
      "48572 2019-11-08 00:07:09 2019-11-08 00:07:09         Wrong Way Driver   \n",
      "48573 2019-11-08 00:06:30 2019-11-08 00:06:30              Person Stop   \n",
      "48574 2019-11-08 00:03:30 2019-11-08 00:08:18                  Dispute   \n",
      "\n",
      "      Disposition  EventNumber                                    Location  \\\n",
      "22034        ADVI     19139370                          342 ADAMS ST, EUG    \n",
      "22035        ADVI     19139367                       3640 E 25TH AVE, EUG    \n",
      "22036        WARN     19139356                    E 10TH AVE/HIGH ST, EUG    \n",
      "22037        PCHK     19139353                   W 4TH AVE/MONROE ST, EUG    \n",
      "22038        GOA      19139347                     5650 BABE RUTH DR, EUG    \n",
      "...           ...          ...                                         ...   \n",
      "48570        ADVI     19291722                  BLAIR BLVD/W 6TH ALY, EUG    \n",
      "48571        ADVI     19291710  GOODPASTURE ISLAND RD/STONE CREST DR, EUG    \n",
      "48572        UTC      19291706                    W 7TH AVE/GRANT ST, EUG    \n",
      "48573        WARN     19291705                    W 5TH AVE/OLIVE ST, EUG    \n",
      "48574        ASST     19291701                       3215 W 11TH AVE, EUG    \n",
      "\n",
      "      Priority Case  \n",
      "22034        3  NaN  \n",
      "22035        3  NaN  \n",
      "22036        6       \n",
      "22037        5  NaN  \n",
      "22038        3  NaN  \n",
      "...        ...  ...  \n",
      "48570        6  NaN  \n",
      "48571        6  NaN  \n",
      "48572        P  NaN  \n",
      "48573        6  NaN  \n",
      "48574        3  NaN  \n",
      "\n",
      "[26541 rows x 8 columns]\n"
     ]
    }
   ],
   "source": [
    "#df.set_index('CallTime')\n",
    "\n",
    "df_june_2020 = df[(df['CallTime']>pd.Timestamp(2019,6,1)) & (df['CallTime']<pd.Timestamp(2019,11,9))]\n",
    "\n",
    "print(df_june_2020)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "RateLimiter caught an error, retrying (0/2 tries). Called with (*('1699 N TERRY ST, Eugene, Oregon ',), **{}).\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/urllib3/connection.py\", line 159, in _new_conn\n",
      "    conn = connection.create_connection(\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/urllib3/util/connection.py\", line 84, in create_connection\n",
      "    raise err\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/urllib3/util/connection.py\", line 74, in create_connection\n",
      "    sock.connect(sa)\n",
      "socket.timeout: timed out\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py\", line 670, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py\", line 381, in _make_request\n",
      "    self._validate_conn(conn)\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py\", line 978, in _validate_conn\n",
      "    conn.connect()\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/urllib3/connection.py\", line 309, in connect\n",
      "    conn = self._new_conn()\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/urllib3/connection.py\", line 164, in _new_conn\n",
      "    raise ConnectTimeoutError(\n",
      "urllib3.exceptions.ConnectTimeoutError: (<urllib3.connection.HTTPSConnection object at 0x7fab38b30cd0>, 'Connection to nominatim.openstreetmap.org timed out. (connect timeout=1)')\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/requests/adapters.py\", line 439, in send\n",
      "    resp = conn.urlopen(\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py\", line 754, in urlopen\n",
      "    return self.urlopen(\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py\", line 754, in urlopen\n",
      "    return self.urlopen(\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py\", line 726, in urlopen\n",
      "    retries = retries.increment(\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/urllib3/util/retry.py\", line 439, in increment\n",
      "    raise MaxRetryError(_pool, url, error or ResponseError(cause))\n",
      "urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=1699+N+TERRY+ST%2C+Eugene%2C+Oregon+&format=json&limit=1 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7fab38b30cd0>, 'Connection to nominatim.openstreetmap.org timed out. (connect timeout=1)'))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/geopy/adapters.py\", line 383, in _request\n",
      "    resp = self.session.get(url, timeout=timeout, headers=headers)\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/requests/sessions.py\", line 543, in get\n",
      "    return self.request('GET', url, **kwargs)\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/requests/sessions.py\", line 530, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/requests/sessions.py\", line 643, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/requests/adapters.py\", line 504, in send\n",
      "    raise ConnectTimeout(e, request=request)\n",
      "requests.exceptions.ConnectTimeout: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=1699+N+TERRY+ST%2C+Eugene%2C+Oregon+&format=json&limit=1 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7fab38b30cd0>, 'Connection to nominatim.openstreetmap.org timed out. (connect timeout=1)'))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/geopy/extra/rate_limiter.py\", line 136, in _retries_gen\n",
      "    yield i  # Run the function.\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/geopy/extra/rate_limiter.py\", line 274, in __call__\n",
      "    res = self.func(*args, **kwargs)\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/geopy/geocoders/nominatim.py\", line 294, in geocode\n",
      "    return self._call_geocoder(url, callback, timeout=timeout)\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/geopy/geocoders/base.py\", line 360, in _call_geocoder\n",
      "    result = self.adapter.get_json(url, timeout=timeout, headers=req_headers)\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/geopy/adapters.py\", line 373, in get_json\n",
      "    resp = self._request(url, timeout=timeout, headers=headers)\n",
      "  File \"/Users/keithbucher/miniconda3/lib/python3.8/site-packages/geopy/adapters.py\", line 395, in _request\n",
      "    raise GeocoderUnavailable(message)\n",
      "geopy.exc.GeocoderUnavailable: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=1699+N+TERRY+ST%2C+Eugene%2C+Oregon+&format=json&limit=1 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7fab38b30cd0>, 'Connection to nominatim.openstreetmap.org timed out. (connect timeout=1)'))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing at 23000 records\n",
      "Writing at 24000 records\n",
      "Writing at 25000 records\n",
      "Writing at 26000 records\n",
      "Writing at 27000 records\n",
      "Writing at 28000 records\n",
      "Writing at 29000 records\n",
      "Writing at 30000 records\n",
      "Writing at 31000 records\n",
      "Writing at 32000 records\n",
      "Writing at 33000 records\n",
      "Writing at 34000 records\n",
      "Writing at 35000 records\n",
      "Writing at 36000 records\n",
      "Writing at 37000 records\n",
      "Writing at 38000 records\n",
      "Writing at 39000 records\n",
      "Writing at 40000 records\n",
      "Writing at 41000 records\n",
      "Writing at 42000 records\n",
      "Writing at 43000 records\n",
      "Writing at 44000 records\n",
      "Writing at 45000 records\n",
      "Writing at 46000 records\n",
      "Writing at 47000 records\n",
      "Writing at 48000 records\n",
      "                 CallTime        DispatchTime             IncidentDesc  \\\n",
      "22034 2019-06-01 23:59:37 2019-06-02 01:20:52        Criminal Trespass   \n",
      "22035 2019-06-01 23:56:05 2019-06-02 00:04:52  Suspicious Condition(s)   \n",
      "22036 2019-06-01 23:46:29 2019-06-01 23:46:29             Traffic Stop   \n",
      "22037 2019-06-01 23:41:23 2019-06-02 04:30:48         Animal Complaint   \n",
      "22038 2019-06-01 23:35:05 2019-06-02 00:17:35  Suspicious Condition(s)   \n",
      "...                   ...                 ...                      ...   \n",
      "48570 2019-11-08 00:45:11 2019-11-08 00:45:11              Person Stop   \n",
      "48571 2019-11-08 00:24:11 2019-11-08 00:24:11              Person Stop   \n",
      "48572 2019-11-08 00:07:09 2019-11-08 00:07:09         Wrong Way Driver   \n",
      "48573 2019-11-08 00:06:30 2019-11-08 00:06:30              Person Stop   \n",
      "48574 2019-11-08 00:03:30 2019-11-08 00:08:18                  Dispute   \n",
      "\n",
      "      Disposition  EventNumber  \\\n",
      "22034        ADVI     19139370   \n",
      "22035        ADVI     19139367   \n",
      "22036        WARN     19139356   \n",
      "22037        PCHK     19139353   \n",
      "22038        GOA      19139347   \n",
      "...           ...          ...   \n",
      "48570        ADVI     19291722   \n",
      "48571        ADVI     19291710   \n",
      "48572        UTC      19291706   \n",
      "48573        WARN     19291705   \n",
      "48574        ASST     19291701   \n",
      "\n",
      "                                                Location Priority Case  \\\n",
      "22034                      342 ADAMS ST, Eugene, Oregon         3  NaN   \n",
      "22035                   3640 E 25TH AVE, Eugene, Oregon         3  NaN   \n",
      "22036                E 10TH AVE/HIGH ST, Eugene, Oregon         6        \n",
      "22037               W 4TH AVE/MONROE ST, Eugene, Oregon         5  NaN   \n",
      "22038                 5650 BABE RUTH DR, Eugene, Oregon         3  NaN   \n",
      "...                                                  ...      ...  ...   \n",
      "48570              BLAIR BLVD/W 6TH ALY, Eugene, Oregon         6  NaN   \n",
      "48571  GOODPASTURE ISLAND RD/STONE CREST DR, Eugene, ...        6  NaN   \n",
      "48572                W 7TH AVE/GRANT ST, Eugene, Oregon         P  NaN   \n",
      "48573                W 5TH AVE/OLIVE ST, Eugene, Oregon         6  NaN   \n",
      "48574                   3215 W 11TH AVE, Eugene, Oregon         3  NaN   \n",
      "\n",
      "             lat         lng  \n",
      "22034  44.056154 -123.106540  \n",
      "22035  44.032117 -123.080951  \n",
      "22036  44.048810 -123.088123  \n",
      "22037  44.055455 -123.104860  \n",
      "22038  44.080973 -123.201732  \n",
      "...          ...         ...  \n",
      "48570  44.053251 -123.107012  \n",
      "48571  44.084017 -123.094993  \n",
      "48572  44.052170 -123.120137  \n",
      "48573  44.054325 -123.094265  \n",
      "48574  44.048638 -123.141189  \n",
      "\n",
      "[26541 rows x 10 columns]\n"
     ]
    }
   ],
   "source": [
    "from geopy.extra.rate_limiter import RateLimiter\n",
    "import time\n",
    "\n",
    "df_short = df_june_2020\n",
    "\n",
    "geocode = RateLimiter(locator.geocode, min_delay_seconds=1)\n",
    "\n",
    "for i, row in df_short.iterrows():    \n",
    "    newLocation = row['Location']\n",
    "    newLocation = newLocation.replace('EUG', 'Eugene, Oregon')\n",
    "    lat = None\n",
    "    lng = None\n",
    "    try:\n",
    "        if '/' in newLocation:\n",
    "            loc = gmaps.geocode(newLocation)\n",
    "            if loc != None  and len(loc) > 0:\n",
    "                lat = loc[0]['geometry']['location']['lat']\n",
    "                lng = loc[0]['geometry']['location']['lng']\n",
    "        else:\n",
    "            loc = geocode(newLocation)\n",
    "            if loc != None:\n",
    "                lat = loc.latitude\n",
    "                lng = loc.longitude                    \n",
    "    except Exception as exc:\n",
    "        print(type(inst))\n",
    "        print(inst.args)\n",
    "        # cool off\n",
    "        time.sleep(60)        \n",
    "    \n",
    "    df_short.at[i,'Location'] = newLocation\n",
    "    df_short.at[i, 'lat'] = lat\n",
    "    df_short.at[i, 'lng'] = lng    \n",
    "    if((i % 1000) == 0):\n",
    "        print('Writing at ' + str(i) + ' records')\n",
    "        df_short.to_csv(\"police_calls_2019_latlon.csv\", index=None, sep=';', encoding='cp1252')\n",
    "    \n",
    "\n",
    "#df_short['Location'] = df_short['Location'].apply(lambda loc: loc.replace('EUG', 'Eugene, Oregon'))\n",
    "#geocode = RateLimiter(locator.geocode, min_delay_seconds=1)\n",
    "#df_short['loc'] = df_short['Location'].apply(geocode)\n",
    "#df_short['gmaps_loc'] = df_short['Location'].apply(lambda loc: gmaps.geocode if '/' in loc else None)\n",
    "#df_short['lat'] = df_short['loc'].apply(lambda loc: loc.latitude if loc else None)\n",
    "#df_short['lng'] = df_short['loc'].apply(lambda loc: loc.longitude if loc else None)\n",
    "\n",
    "df_short.to_csv(\"police_calls_2019_latlon.csv\", index=None, sep=';', encoding='cp1252')\n",
    "\n",
    "print(df_short)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
